### ISSUE TITLE
Certain Reports Require RDT by July 31, 2023,

### ISSUE THREAD

**Issue by zoborg** [jlevers/selling-partner-api#262]
Firstly, thanks for this library - great work, much appreciated!

**Current Behaviour**

When requesting reports e.g

$apiInstance->getReportDocument($report_document_id, $report_type);

No RDT token is sent in the request

**Expected Behaviour** 

If a report is deemed to require a RDT Token, this should be set when  getting Report Document

**Rationale**

As per a recent update from Amazon, certain reports will require RDT, see below

Important Notice: Discontinued Support - LWA Tokens can no longer retrieve PII in Selling Partner API

What is changing?

Effective July 31, 2023, Developer applications must use the new access protocol via the Tokens API and Restricted Data Token (RDT) to make successful calls to retrieve Personally Identifiable Information (PII) in the Selling Partner API. Use of LWA tokens to try and retrieve this information will result in a failure. The operations listed below will require the use of RDT and will no longer accept LWA tokens. In addition, all operations and reports that vend PII, launched after July 31, 2023, will require the use of RDT and will not support LW tokens at launch.

Orders
GetOrderAddress
GetOrderItemsBuyerInfo
GetOrders
GetOrder
GetOrderItems
Merchant Fulfillment
GetShipments
CreateShipment
CancelShipmentOld
Shipping
directPurchaseShipment
Direct Fulfillment Orders
GetOrder
GetOrders
Direct Fulfillment Shipping
GetCustomerInvoices
GetCustomerInvoice
GetPackingSlips
GetPackingSlip
GetShippingLabels
GetShippingLabel
Reports
GetReportDocument
GET_AMAZON_FULFILLED_SHIPMENTS_DATA_INVOICING
GET_AMAZON_FULFILLED_SHIPMENTS_DATA_TAX
GET_FLAT_FILE_ACTIONABLE_ORDER_DATA_SHIPPING
GET_FLAT_FILE_ORDER_REPORT_DATA_SHIPPING
GET_FLAT_FILE_ORDER_REPORT_DATA_INVOICING
GET_FLAT_FILE_ORDER_REPORT_DATA_TAX
GET_FLAT_FILE_ORDERS_RECONCILIATION_DATA_TAX
GET_FLAT_FILE_ORDERS_RECONCILIATION_DATA_INVOICING
GET_FLAT_FILE_ORDERS_RECONCILIATION_DATA_SHIPPING
GET_ORDER_REPORT_DATA_INVOICING
GET_ORDER_REPORT_DATA_TAX
GET_ORDER_REPORT_DATA_SHIPPING
GET_EASYSHIP_DOCUMENTS
GET_GST_MTR_B2B_CUSTOM
GET_VAT_TRANSACTION_DATA
SC_VAT_TAX_REPORT
Why is this changing?

Amazon takes protecting customer data seriously. This change ensures that developers are using the latest technology and features for the benefit of vendors, sellers, and customers.

Who is affected?

Developers who use Selling Partner APIs to retrieve customer PII.

Which countries are affected?

This affects all countries.

What action is required?

To avoid service disruption, you must update affected applications to use the new access protocol by July 31, 2023.

**Comment by jlevers** [comment]
Thanks for pointing this out -- I'll add support for this into v5.0, which should be coming out in the next week or two.

**Comment by zoborg** [comment]
@jlevers - you legend! thank you

**Comment by tazplan** [comment]
Was it the reason why we got  GET_FLAT_FILE_ACTIONABLE_ORDER_DATA_SHIPPING status systematically CANCELLED with older version?

**Comment by jlevers** [comment]
@zoborg I finally got around to looking at this, and I'm pretty sure this is already handled right here:

https://github.com/jlevers/selling-partner-api/blob/dd59c468aed46c1a5623b882898716c9ed269147/lib/Authentication.php#L173-L185

Do you have some reason to believe that this wasn't working for you?

**Comment by zoborg** [comment]
@jlevers 

Unless I've misunderstood it doesn't get to that section, as per the documentation (https://github.com/jlevers/selling-partner-api#downloading-a-report-document)

            $reportType = ReportType::GET_ORDER_REPORT_DATA_TAX;
            $reportsApi = new ReportsApi($config);
            $reportDocumentInfo = $reportsApi->getReportDocument($documentId, $reportType['name']);

Would be an example request, when signRequest is called scope and the restricted path is null (and dataElements empty) so it always hits the first if condition.

        if (!$this->signingScope && ($restrictedPath === null || $dataElements === [])) {
            $relevantCreds = $this->getAwsCredentials();
        }

**Comment by zoborg** [comment]
@jlevers - did you have any thoughts on this? the deadline is dooming?

**Comment by jlevers** [comment]
@zoborg just to explain the fix:

> signRequest is called scope and the restricted path is null (and dataElements empty)

when you call `getReportDocument`, `$this->signingScope` will be `null`, and `$restrictedPath` will be the report document retrieval path (e.g. `/reports/2021-06-30/documents/amzn1.spdoc.1.3.5........R66.129`), per this section of code:

https://github.com/jlevers/selling-partner-api/blob/8b5cdc6739242bc87b2445070cc8165569381765/lib/Api/ReportsV20210630Api.php#L1907-L1912

the fix was to switch this line that you highlighted:

https://github.com/jlevers/selling-partner-api/blob/8b5cdc6739242bc87b2445070cc8165569381765/lib/Authentication.php#L165

to

```php
if (!$this->signingScope && $restrictedPath === null && $dataElements === []) { 
```

and then the RDT gets passed to the API as expected.

**Comment by zoborg** [comment]
@jlevers - perfect works a treat!

**Comment by jlevers** [comment]
Fixed in `v5.2.0`.
